baseServer_test.go

```diff
@@ -0,0 +1,377 @@
+package basePlatformSOMAS_test
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/MattSScott/basePlatformSOMAS"
+	"github.com/google/uuid"
+)
+
+type ITestBaseAgent interface {
+	basePlatformSOMAS.IAgent[ITestBaseAgent]
+	NewTestMessage() TestMessage
+	HandleTestMessage()
+	ReceivedMessage() bool
+	GetCounter() int
+	SetCounter(int)
+	GetGoal() int
+	SetGoal(int)
+}
+
+type ITestServer interface {
+	basePlatformSOMAS.IServer[ITestBaseAgent]
+	basePlatformSOMAS.PrivateServerFields[ITestBaseAgent]
+}
+
+type TestAgent struct {
+	*basePlatformSOMAS.BaseAgent[ITestBaseAgent]
+	counter int
+	goal    int
+	mu      *sync.Mutex
+}
+
+type TestServer struct {
+	*basePlatformSOMAS.BaseServer[ITestBaseAgent]
+	roundCounter int
+}
+
+type TestMessage struct {
+	basePlatformSOMAS.BaseMessage
+	counter int
+}
+
+func (tm TestMessage) InvokeMessageHandler(ag ITestBaseAgent) {
+	ag.HandleTestMessage()
+}
+
+func (tba *TestAgent) CreateTestMessage() TestMessage {
+	return TestMessage{
+		basePlatformSOMAS.BaseMessage{},
+		5,
+	}
+}
+
+func (tba *TestAgent) SetCounter(count int) {
+	tba.counter = count
+}
+
+func (tba *TestAgent) SetGoal(goal int) {
+	tba.goal = goal
+}
+func (tba *TestAgent) GetGoal() int {
+	return tba.goal
+}
+func NewTestMessage() TestMessage {
+	return TestMessage{
+		basePlatformSOMAS.BaseMessage{},
+		5,
+	}
+}
+
+func NewTestServer(generatorArray []basePlatformSOMAS.AgentGeneratorCountPair[ITestBaseAgent], iterations, turns int, maxDuration time.Duration) *TestServer {
+	return &TestServer{
+		BaseServer:   basePlatformSOMAS.CreateServer(generatorArray, iterations, turns, maxDuration),
+		roundCounter: 0,
+	}
+}
+
+func NewTestAgent(serv basePlatformSOMAS.IExposedServerFunctions[ITestBaseAgent]) ITestBaseAgent {
+	mu := &sync.Mutex{}
+	return &TestAgent{
+		BaseAgent: basePlatformSOMAS.CreateBaseAgent(serv),
+		counter:   0,
+		goal:      0,
+		mu:        mu,
+	}
+}
+
+func (ta2 *TestAgent) NewTestMessage() TestMessage {
+	return TestMessage{
+		basePlatformSOMAS.CreateBaseMessage(ta2.GetID()),
```

Comment:
I'm actually tempted to move `CreateBaseMessage` into BaseAgent. That way, it can be called without having to manually pass the sender ID. I.e., we handle the base agent implementation (which calls the method from messaging) and then just call this from the composing class